### PR TITLE
Guard to call "close" after "get", "put", "del" and "batch"

### PR DESCRIPTION
**Pull request by braydonf**

### CODE REVIEW

**Comment by braydonf** [comment]
Updated from feedback.

**Comment by braydonf** [comment]
`guardCallback` has been updated

**Comment by jcrugzz** [comment]
Not a bad thing to have to ensure safety, I'd be curious if it causes any perf issues but it is only 1 extra function call

**Comment by braydonf** [comment]
Running the benchmarks with release as levelup@1.0.0-5 and leveldown@1.0.6:
```
LevelUP           put(int, string) x 10 x 659 ops/sec ±1.40% (44 runs sampled)
LevelUP (release) put(int, string) x 10 x 647 ops/sec ±1.02% (77 runs sampled)

LevelUP           put(int, string) x 1000 x 136 ops/sec ±3.70% (68 runs sampled)
LevelUP (release) put(int, string) x 1000 x 161 ops/sec ±3.15% (81 runs sampled)

LevelUP           put(int, string) x 100,000 x 1.73 ops/sec ±5.15% (13 runs sampled)
LevelUP (release) put(int, string) x 100,000 x 1.70 ops/sec ±9.16% (12 runs sampled)

LevelUP           get(int):string x 10 x 664 ops/sec ±1.97% (37 runs sampled)
LevelUP (release) get(int):string x 10 x 664 ops/sec ±2.04% (41 runs sampled)

LevelUP           get(int):string x 1000 x 124 ops/sec ±0.76% (69 runs sampled)
LevelUP (release) get(int):string x 1000 x 131 ops/sec ±0.90% (79 runs sampled)

LevelUP           batch(int, string) x 1000 x 201 ops/sec ±3.63% (60 runs sampled)
LevelUP (release) batch(int, string) x 1000 x 209 ops/sec ±3.21% (62 runs sampled)

LevelUP           batch(int, string) x 100,000 x 2.60 ops/sec ±2.28% (17 runs sampled)
LevelUP (release) batch(int, string) x 100,000 x 2.69 ops/sec ±0.84% (18 runs sampled)
```

**Comment by braydonf** [comment]
Updated. BTW: jshint isn't giving any warnings about the style changes, should it?

**Comment by jcrugzz** [comment]
@ralphtheninja we should think about making a `level-styleguide` with something like https://www.npmjs.com/package/fashion-show

**Comment by ralphtheninja** [comment]
@jcrugzz Awesome idea. Feel like hacking together one? :)
@braydonf It should :)

File: lib/levelup.js [hunk 1]
```diff
@@ -132,7 +129,9 @@ LevelUP.prototype.open = function (callback) {
 LevelUP.prototype.close = function (callback) {
   var self = this
 
-  if (this.isOpen()) {
+  if (this._activeOperations > 0) {
+    this.on('inactive', this.close.bind(this, callback))
```
Comment (juliangruber) [inline lib/levelup.js]: should be a `.once`, because you can close and open a db more than once

**Comment by juliangruber** [comment]
for a complete solution this needs to guard against creation of new streams as well

**Comment by juliangruber** [comment]
since leveldown takes care of closing iterators when you `.close()`, maybe it should take care of the async functions as well. see https://github.com/Level/leveldown/blob/master/test/cleanup-hanging-iterators-test.js#L32

hmmm my gut tells me this should all either go into levelup or abstract-leveldown, so none of the backends needs to take care of this stuff. what do you think?

File: lib/levelup.js (hunk 1 above)
Comment (ralphtheninja) [inline lib/levelup.js]: :+1:

**Comment by braydonf** [comment]
Initially I thought the best place would be in `Database::Close` in leveldown, similar to the [iterators check](https://github.com/Level/leveldown/blob/master/src/database.cc#L263), however after running gdb with levelup code similar to:
```js
db.put('key', 'value', function() {});
db.close(function() {});
```
`Database::Close` was called and completed *first*, even though put was called. So having the guard as close to the caller and the original order makes sense.

Though I still think the backends should handle it to some degree, i.e. in the case of leveldown, if `Database::Put` is called after `Database::Close` it should [give an error](https://github.com/Level/leveldown/issues/157#issuecomment-100020704) rather than crash with a segfault.

**Comment by ralphtheninja** [comment]
Closing due to outdated and too much stuff has happened lately to be able to merge this.

**Pull request closed without merging**